src/diepvries/template_sql/effectivity_satellite_dml.sql

```diff
@@ -1,78 +1,84 @@
-        SELECT
-          {staging_driving_keys},
-          staging.{hashkey_field},
-          staging.{staging_hashdiff_field},
-          staging.{record_start_timestamp},
-          staging.{record_source}
-          {staging_descriptive_fields}
-        FROM filtered_staging AS staging
-          LEFT JOIN effectivity_satellite AS satellite
-                    ON ({satellite_driving_key_condition})
-        WHERE satellite.{hashkey_field} IS NULL
-           OR satellite.{hashdiff_field} <> staging.{staging_hashdiff_field}
-        UNION ALL
-        /* Records from the target table that will have its r_timestamp_end updated
-          (hashkey already exists in target table, but hashdiff changed). As the
-          r_timestamp is fetched from the target table, these records will always be
-          included in the WHEN MATCHED condition of the MERGE command. */
+                                   ),
+          filtered_effectivity_satellite AS (
+          SELECT
+            satellite.*
+          FROM {staging_schema}.{staging_table} AS staging
+            INNER JOIN effectivity_satellite AS satellite
+                       ON ({satellite_driving_key_condition})
+                                            ),
+          filtered_staging AS (
+          SELECT DISTINCT
+            {staging_driving_keys},
+            staging.{hashkey_field},
+            staging.{staging_hashdiff_field},
+            staging.{record_start_timestamp},
+            staging.{record_source}
+            {staging_descriptive_fields}
+          FROM {staging_schema}.{staging_table} AS staging
+          WHERE NOT EXISTS (
+                           SELECT
+                             1
+                           FROM filtered_effectivity_satellite AS satellite
+                           WHERE {satellite_driving_key_condition}
+                            AND satellite.r_timestamp >= staging.r_timestamp
+                           )
+                              ),
+          --   Records that will be inserted (don't exist in target table or exist
+          --   in the target table but the hashdiff changed). As the r_timestamp is fetched
+          --   from the staging table, these records will always be included in the
+          --   WHEN NOT MATCHED condition of the MERGE command.
+          staging_satellite_affected_records AS (
+          SELECT
+            {staging_driving_keys},
+            staging.{hashkey_field},
+            staging.{staging_hashdiff_field},
+            staging.{record_start_timestamp},
+            staging.{record_source}
+            {staging_descriptive_fields}
+          FROM filtered_staging AS staging
+            LEFT JOIN filtered_effectivity_satellite AS satellite
+                      ON ({satellite_driving_key_condition})
+          WHERE satellite.{hashkey_field} IS NULL
+             OR satellite.{hashdiff_field} <> staging.{staging_hashdiff_field}
```

Comment:
Why do we have this OR condition here? Couldn't that get us records that are already in the satellite?

Reply:
Hum, good question! In the end the records that are caught by this condition should actually be caught in the next `UNION ALL` query, as they exist in the target table hence should be returned by the `INNER JOIN` and caught by the condition ` WHERE satellite.{hashdiff_field} <> staging.{staging_hashdiff_field}`.

We never had issues with it, but we should have right?

Reply:
Issues like duplicates?

Reply:
Ah I think I remember. The difference here is that in the first union all query we're fetching the fields from the staging.<field_name> and the second we fetch the fields from the satellite.<field_name>.

So in the first union all we want either the new version of existing records (source from stg, different hashdiff) or the first version of records that never existed!

Whereas in the next one we want to fetch the existing version from the sateliite with new versions on the staging table.

The key here is that the staging.{record_start_timestamp} will always be different than the satellite.{record_start_timestamp} so these will always be different versions and both should be considered in the query as we do a merge by primary key (hashkey + r_timestamp or driving key + r_timestamp)

Does this make sense?

Reply:
@tsferreira4 I already have the answer (above), but I just answered it in the wrong thread!

There will be no duplicates because each of the UNION ALL queries fetch fields from a different table. THe first one fetches fields from the staging table whereas the second one fetches fields from the satellite, hence they're effectively different records.

Reply:
But yes, otherwise we'd have issues with duplicates.

Reply:
Makes sense! Maybe we could improve a bit the docs? Now I see it very clearly, but maybe in a year we'll ask ourselves the same

Reply:
Sure, will do

Reply:
Fair enough!